src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -92,13 +129,30 @@ private boolean isNewTimestamp() {
         return true;
     }
 
-    private Pose3d doubleArrayToPose3d(double[] doubleArray) {
-        if (doubleArray == null || doubleArray.length != 6)
-            return null;
+    private void logVisibleTags() {
+        if (inputs.hasResult) {
+            Logger.recordOutput("VisibleTags/" + this.getName(), new Pose2d[0]);
+            return;
+        }
 
-        return new Pose3d(
-                new Translation3d(doubleArray[0], doubleArray[1], doubleArray[2]),
-                new Rotation3d(doubleArray[3], doubleArray[4], doubleArray[5])
-        );
+        final Pose2d[] visibleTagPoses = new Pose2d[inputs.visibleTagIDs.length];
+        for (int i = 0; i < visibleTagPoses.length; i++)
+            visibleTagPoses[i] = FieldConstants.TAG_ID_TO_POSE.get(i).toPose2d();
+        Logger.recordOutput("VisibleTags/" + this.getName(), visibleTagPoses);
+    }
+
+    private Matrix<N3, N1> solvePNPAverageDistanceToStdDevs() {
+        final int numberOfVisibleTags = inputs.visibleTagIDs.length;
+        final double translationStd = translationsStdExponent * Math.pow(inputs.averageDistanceFromAllTags, 2) / (numberOfVisibleTags * numberOfVisibleTags);
+        final double thetaStd = thetaStdExponent * Math.pow(inputs.averageDistanceFromAllTags, 2) / numberOfVisibleTags;
+
+        return VecBuilder.fill(translationStd, translationStd, thetaStd);
+    }
+
+    private Matrix<N3, N1> assumedHeadingAverageDistanceToStdDevs() {
```

Comment:
The distance ain't average
Just call this calculateAssumedHeadingStdDevs

Add a javadoc that explains why the theta std dev is inf, and explain generally how this works

Reply:
I don't know how it works

Reply:
Well, then fix that
Read [this pose](https://www.chiefdelphi.com/t/frc-6328-mechanical-advantage-2023-build-thread/420691/36?u=levyishai), specially from `Localization & Pose Ambiguity` (reading the entire post is also good, it's very interesting)

Reply:
https://www.chiefdelphi.com/t/frc-6328-mechanical-advantage-2023-build-thread/420691/292?u=levyishai this as well